src/software/replay/replay_test.cpp

```diff
@@ -0,0 +1,88 @@
+#include <google/protobuf/util/message_differencer.h>
+#include <gtest/gtest.h>
+
+#include "software/multithreading/subject.h"
+#include "software/replay/replay_logger.h"
+#include "software/replay/replay_reader.h"
+
+constexpr const char* REPLAY_TEST_PATH_SUFFIX = "software/replay/test_replay";
+
+namespace fs = std::experimental::filesystem;
+
+// hahaha test subject because it's a subject but also a test subject
+class TestSubject : public Subject<SensorMsg>
+{
+   public:
+    void sendValue(SensorMsg val)
+    {
+        sendValueToObservers(val);
+    }
+};
+
+TEST(ReplayTest, test_read_and_write_replay)
+{
+    // unfortunately due to the unavailablity of ordering tests and functions that
+    // generate test files that actually work, we have to read recorded replays back,
+    // write them, and then read them again in the same test
+    // TODO: record replay data with refbox running
+
+    std::vector<SensorMsg> read_replay_msgs;
+
+    ReplayReader reader(fs::current_path() / REPLAY_TEST_PATH_SUFFIX);
+    while (auto frame = reader.getNextMsg())
+    {
+        read_replay_msgs.emplace_back(*frame);
+    }
+    // do a quick sanity check here to assert that the timestamps are monotonically
+    // increasing
+    double max_timestamp = 0.f;
+    for (const auto& msg : read_replay_msgs)
+    {
+        auto ts = msg.ssl_vision_msg().detection().t_sent();
+        if (ts < max_timestamp)
+        {
+            FAIL() << "t_sent " << ts << " should have appeared after " << max_timestamp;
+        }
+        else
+        {
+            max_timestamp = ts;
+        }
+    }
+
+    // write the read frames to another replay directory
+    auto output_path = fs::current_path() / "replaytest";
+    std::shared_ptr<Observer<SensorMsg>> logger_ptr =
+        std::make_shared<ReplayLogger>(output_path);
+    TestSubject subject;
+    subject.registerObserver(logger_ptr);
+    for (const auto msg : read_replay_msgs)
+    {
+        subject.sendValue(msg);
+        // we have 12000 frames of replay so we have to try to not fill the buffer
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
nit: this will cause the test to take at least 12 seconds (which is fine, but there might be an easy optimization here). WOuld it be sufficient to log fewer frames but make the chunks smaller? Then you can still test the chunk system but don't need so much data

Reply:
can probably delete most of the chunks in the folder to speed up the test